### ISSUE TITLE
Bug: SA1306 (FieldNamesMustBeginWithLowerCaseLetter) not reported for local variables

### ISSUE THREAD

**Issue by test20150828** [DotNetAnalyzers/StyleCopAnalyzers#1307]
Old StyleCop reports a warning when a local variable name begins with an upper-case letter. StyleCop.Analyzers (1.0.0.8) check field names only:

```C#
public class Class1
{
	public void TestSA1306()
	{
		// no warning here
		int SA1306shouldBeReportedHere = 0;
		++SA1306shouldBeReportedHere;
	}
}
```

**Comment by sharwell** [comment]
Definitely interesting. My thoughts:

* It doesn't make sense to report SA1306 here. A local variable is not a field.

* I'm not sure it matters enough to check this case. We could resurrect SA1301 (ElementMustBeginWithLowerCaseLetter), but my vote is to keep the current behavior and document it as a known deviation from StyleCop "classic".

**Comment by test20150828** [comment]
The documentation says that *A violation of this rule occurs when the name of a field or variable begins with an upper-case letter*. After switching from old StyleCop to StyleCop.Analyzer we have no rule to ensure that nobody writes code like this:
```C#
void Test()
{
    string URL = "http://server"; // should be 'url'
    int[] Data = new int[] { 1, 2, 3 }; // should be 'data'
}
```
If extending SA1306 is not acceptable, then adding a new similar rule for locals would be very desirable.

**Comment by sharwell** [comment]
@oatkins What do you think here?

**Comment by pdelvo** [comment]
I would do one of these three things:

* Implement variable analysis in SA1306 and change the name of the diagnostic to include that. Maybe it would be a good idea to make it configurable
* Don't include it in SA1306, remove it from the documentation, and instead add a new rule for variables
* Report it as SA1301 and change the documentation to reflect that

I personally like option 2 the most

**Comment by otac0n** [comment]
I would vote for number 1, `VariablesMustBeginWithALowerCaseLetter`.  To me, the original intent of the rule was to provide a baseline for all variables, fields included, which can then be overridden by other rules.

PS: @test20150828 Did you realize you just opened bug #1306 referencing SA1307, and you opened bug #1307 referencing SA1306?  I thought it was a neat coincidence.

**Comment by Noryoko** [comment]
I would prefer a new rule be added for local variables. Gives the user the option for checking fields, locals, or both without the need to implement configuration options.

**Comment by sharwell** [comment]
I filed new rule proposals #1310 and #1311 based on the responses here.

Thanks for the feedback @test20150828! I expect these to be accepted and implemented quickly.
